engine/src/main/java/io/zeebe/engine/nwe/WorkflowInstanceStateTransitionGuard.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.engine.nwe;
+
+import io.zeebe.engine.Loggers;
+import io.zeebe.engine.nwe.behavior.BpmnStateBehavior;
+import io.zeebe.engine.state.instance.ElementInstance;
+import io.zeebe.protocol.record.intent.WorkflowInstanceIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import io.zeebe.util.Either;
+import org.slf4j.Logger;
+
+/**
+ * A check to prevent concurrent state transitions of a workflow instance.
+ *
+ * <p>A workflow instance can be have concurrent state transitions if a user command is received
+ * (e.g. cancel workflow instance) or if an internal/external event is triggered (e.g. timer
+ * boundary event). In this case, the current workflow instance processing needs to be interrupted
+ * be avoid an inconsistent state.
+ */
+public final class WorkflowInstanceStateTransitionGuard {
+
+  private static final Logger LOGGER = Loggers.WORKFLOW_PROCESSOR_LOGGER;
+
+  private final BpmnStateBehavior stateBehavior;
+
+  public WorkflowInstanceStateTransitionGuard(final BpmnStateBehavior stateBehavior) {
+    this.stateBehavior = stateBehavior;
+  }
+
+  /**
+   * Checks if a workflow instance event can be processed based on the current state.
+   *
+   * @return {@code true} if the transition is valid.
+   */
+  public boolean isValidStateTransition(final BpmnElementContext context) {
+    final var result = checkStateTransition(context);
+
+    // log the reason for better debugging
+    result.ifLeft(
+        violation ->
+            LOGGER.debug(
+                "Don't process event because of an illegal state transition: {} [context: {}]",
+                violation,
+                context));
+
+    return result.isRight();
+  }
+
+  private Either<String, ?> checkStateTransition(final BpmnElementContext context) {
+    switch (context.getIntent()) {
+      case ELEMENT_ACTIVATING:
+      case ELEMENT_ACTIVATED:
+      case ELEMENT_COMPLETING:
+      case ELEMENT_COMPLETED:
+        return hasElementInstanceWithState(context, context.getIntent())
+            .flatMap(ok -> hasActiveFlowScopeInstance(context));
```

Comment:
shouldn't we check the `ok` value?

Reply:
Nope. `ok` is just ok :stuck_out_tongue: 
In this case, `ok` is from type `Void` and we don't want to consume it.